server/server.go

```diff
@@ -1068,6 +1071,77 @@ func writeHealthResponse(w http.ResponseWriter, err error, payload interface{})
 	writer.JSON(w, status, payload, false)
 }
 
+func (s *Server) unversionedGetHealthWithPolicy(w http.ResponseWriter, r *http.Request) {
+	pluginStatus := s.manager.PluginStatus()
+	pluginState := map[string]string{}
+
+	// optimistically assume all plugins are ok
+	allPluginsOk := true
+	// iterate over plugin status to extract state
+	for name, status := range pluginStatus {
+		if status != nil {
+			pluginState[name] = string(status.State)
+			// if all plugins have not been in OK state yet, then check to see if plugin state is OKx
+			if !s.allPluginsOkOnce && status.State != plugins.StateOK {
+				allPluginsOk = false
+			}
+		}
+	}
+	// once all plugins are OK, set the allPluginsOkOnce flag to true, indicating that all
+	// plugins have achieved a "ready" state at least once on the server.
+	if allPluginsOk {
+		s.allPluginsOkOnce = true
+	}
+
+	input := map[string]interface{}{
+		"plugin_state":  pluginState,
+		"plugins_ready": s.allPluginsOkOnce,
+	}
+
+	vars := mux.Vars(r)
+	urlPath := vars["path"]
+	healthDataPath := fmt.Sprintf("/system/health/%s", urlPath)
+	healthDataPath = stringPathToDataRef(healthDataPath).String()
+
+	rego := rego.New(
+		rego.Query(healthDataPath),
+		rego.Compiler(s.getCompiler()),
+		rego.Store(s.store),
+		rego.Input(input),
+		rego.Runtime(s.runtime),
+	)
+
+	rs, err := rego.Eval(r.Context())
+
+	if err != nil {
+		writeHealthPolicyResponse(w, err)
+		return
+	}
+
+	if len(rs) == 0 {
+		writeHealthPolicyResponse(w, fmt.Errorf("Health policy failed at %s", healthDataPath))
+		return
+	}
+
+	result, ok := rs[0].Expressions[0].Value.(bool)
+
+	if ok && result {
+		writeHealthPolicyResponse(w, nil)
+		return
+	}
+
+	writeHealthPolicyResponse(w, fmt.Errorf("Health policy did not evaluate to true at %s", healthDataPath))
+}
+
+func writeHealthPolicyResponse(w http.ResponseWriter, err error) {
+	status := http.StatusOK
+	if err != nil {
+		status = http.StatusInternalServerError
+		logrus.WithField("err", err).Error("Health check failed")
```

Comment:
nit: can we avoid this logging call here? The access logs on the server will report this (i.e., "500 internal error on /health/foo") albeit someone less explicitly. I forget if the error text is included in the server log. We could solve that separately if need be. The reason is that the server doesn't have any other logging calls in it today.

Reply:
I'm cool with dropping the error logging from server as long as it's ok if we provide the same verbosity in error message somewhere else, either in the response body from the health route, or in the access logs (which don't show the error message today). Thoughts on which is the better approach?

Reply:
I'd prefer to include the message in the response body. It looks like we currently write an empty object (`{}`) in all cases. Why don't we:

* return an empty object in the healthy case (w/ HTTP 200)
* return an object like `{"error": <msg>}` w/ HTTP 500 in the unhealthy case

It looks like we already declare a few error messages so those'll be visible to the caller now.

Reply:
Should I change the existing `/health` route to return error messages also?

Reply:
I assumed yes.  It's a breaking interface change, but it makes the original health endpoint more useful.

Reply:
Sorry for the delayed response. This is fine, it's not really breaking because we're just adding a new field to the response (which callers shouldn't be affected by).